pkg/activator/revision_backends.go

```diff
@@ -0,0 +1,284 @@
+	transport      http.RoundTripper
+	logger         *zap.SugaredLogger
+	probeFrequency time.Duration
+}
+
+// NewRevisionBackendsManagerWithProbeFrequency returnes a RevisionBackendsManager that uses the supplied
+// probe frequency
+func NewRevisionBackendsManagerWithProbeFrequency(updateCh chan<- *RevisionDestsUpdate,
+	transport http.RoundTripper, endpointsInformer corev1informers.EndpointsInformer,
+	logger *zap.SugaredLogger, probeFrequency time.Duration) *RevisionBackendsManager {
+	rbm := &RevisionBackendsManager{
+		revisionWatchers: make(map[RevisionID]*revisionWatcherCh),
+		updateCh:         updateCh,
+		transport:        transport,
+		logger:           logger,
+		probeFrequency:   probeFrequency,
+	}
+
+	endpointsInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: reconciler.ChainFilterFuncs(
+			reconciler.LabelExistsFilterFunc(serving.RevisionUID),
+			// We are only interested in the private services, since that is
+			// what is populated by the actual revision backends.
+			reconciler.LabelFilterFunc(networking.ServiceTypeKey, string(networking.ServiceTypePrivate), false),
+		),
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    rbm.endpointsUpdated,
+			UpdateFunc: controller.PassNew(rbm.endpointsUpdated),
+			DeleteFunc: rbm.endpointsDeleted,
+		},
+	})
+
+	return rbm
+}
+
+func NewRevisionBackendsManager(updateCh chan<- *RevisionDestsUpdate,
+	transport http.RoundTripper, endpointsInformer corev1informers.EndpointsInformer,
+	logger *zap.SugaredLogger) *RevisionBackendsManager {
+	return NewRevisionBackendsManagerWithProbeFrequency(updateCh, transport, endpointsInformer,
+		logger, probeFrequency)
+}
+
+func (rbm *RevisionBackendsManager) getOrCreateDestsCh(rev RevisionID) chan []string {
+	rbm.revisionWatchersMux.Lock()
+	defer rbm.revisionWatchersMux.Unlock()
+
+	rwCh, ok := rbm.revisionWatchers[rev]
+	if !ok {
+		destsCh := make(chan []string)
+		rw := newRevisionWatcher(rev, rbm.updateCh, destsCh, rbm.transport, rbm.logger)
+		rbm.revisionWatchers[rev] = &revisionWatcherCh{rw, destsCh}
+		go rw.run(rbm.probeFrequency)
+		return destsCh
+	}
+
+	return rwCh.ch
+}
+
+// endpointsUpdated is a handler function to be used by the Endpoints informer.
+// It updates the endpoints in the RevisionBackendsManager if the hosts changed
+//
+// This function must not be called in parallel to not induce a wrong order of events.
+func (rbm *RevisionBackendsManager) endpointsUpdated(newObj interface{}) {
+	endpoints := newObj.(*corev1.Endpoints)
+	revID := RevisionID{endpoints.Namespace, endpoints.Labels[serving.RevisionLabelKey]}
+
+	rbm.getOrCreateDestsCh(revID) <- endpointsToDests(endpoints)
+}
+
+// deleteRevisionWatcher deletes the revision wathcher for rev if it exists. It expects
+// a write lock is held on revisoinWatchersMux when calling.
+func (rbm *RevisionBackendsManager) deleteRevisionWatcher(rev RevisionID) {
+	if rw, ok := rbm.revisionWatchers[rev]; ok {
+		close(rw.ch)
```

Comment:
this feels like layer violation.
Probably better to have `rw.shutdown()` method that will deal with individual revision watcher termination.

Reply:
The revisionWatcher only gets readonly access to the dests ch - IMO this is better for least access principle